runtime/coverage.go

```diff
@@ -20,56 +20,228 @@ package runtime
+		LineHits:   lineHits,
+		Statements: len(lineHits),
 	}
 }
 
-// CoverageReport is a collection of coverage per location
+// CoverageReport collects coverage information per location.
+// It keeps track of inspected programs per location, and can
+// also exclude locations from coverage collection.
 type CoverageReport struct {
+	// Contains a *LocationCoverage per location.
 	Coverage map[common.Location]*LocationCoverage `json:"-"`
+	// Contains an *ast.Program per location.
+	Programs map[common.Location]*ast.Program `json:"-"`
+	// Contains locations excluded from coverage collection.
+	ExcludedLocations []common.Location `json:"-"`
 }
 
+// ExcludeLocation adds the given location to the array of excluded
+// locations.
+func (r *CoverageReport) ExcludeLocation(location Location) {
+	if r.IsLocationExcluded(location) {
+		return
+	}
+	r.ExcludedLocations = append(r.ExcludedLocations, location)
+}
+
+// IsLocationExcluded checks whether the given location is excluded
+// or not, from coverage collection.
+func (r *CoverageReport) IsLocationExcluded(location Location) bool {
+	for _, excludedLocation := range r.ExcludedLocations {
+		if excludedLocation == location {
+			return true
+		}
+	}
+	return false
+}
+
+// AddLineHit increments the hit count for the given line, on the given
+// location. The method call is a NO-OP in two cases:
+// - If the location is excluded from coverage collection
+// - If the location's *ast.Program, has not been inspected
 func (r *CoverageReport) AddLineHit(location Location, line int) {
-	locationCoverage := r.Coverage[location]
-	if locationCoverage == nil {
-		locationCoverage = NewLocationCoverage()
-		r.Coverage[location] = locationCoverage
+	if r.IsLocationExcluded(location) {
+		return
+	}
+
+	if !r.IsProgramInspected(location) {
+		return
 	}
+
+	locationCoverage := r.Coverage[location]
 	locationCoverage.AddLineHit(line)
 }
 
+// InspectProgram inspects the elements of the given *ast.Program, and counts its
+// statements. If inspection is successful, the *ast.Program is marked as inspected.
+// If the given location is excluded from coverage collection, the method call
+// results in a NO-OP.
+func (r *CoverageReport) InspectProgram(location Location, program *ast.Program) {
+	if r.IsLocationExcluded(location) {
+		return
+	}
+	r.Programs[location] = program
+	lineHits := make(map[int]int, 0)
+	var depth int
+
+	inspector := ast.NewInspector(program)
+	inspector.Elements(
+		nil, func(element ast.Element, push bool) bool {
+			if push {
+				depth++
+
+				_, isStatement := element.(ast.Statement)
+				_, isDeclaration := element.(ast.Declaration)
+				_, isVariableDeclaration := element.(*ast.VariableDeclaration)
+
+				// Track only the statements that are not declarations, such as:
+				// - *ast.CompositeDeclaration
+				// - *ast.SpecialFunctionDeclaration
+				// - *ast.FunctionDeclaration
+				if (isStatement && !isDeclaration) ||
+					(isVariableDeclaration && depth > 2) {
```

Comment:
```suggestion
				// - *ast.FunctionDeclaration
				// However, also track local (i.e. non-top level) variable declarations
				if (isStatement && !isDeclaration) ||
					(isVariableDeclaration && depth > 2) {

```

Reply:
Missed that one :sweat_smile: Added in https://github.com/onflow/cadence/pull/2296/commits/84b077202f093176e82931cffe2d1b2e19f0246a#diff-dd50d117359d58b9ec20ba8e56dc4b6b224a84ca819d943436617baec41d34d5R169